### PR TITLE
Use Windows high contrast black/white theme with `MaterialApp` themes

### PR DESCRIPTION
**Pull request by yaakovschectman**
`MaterialApp` has a `highContrastTheme` and a `highContrastDarkTheme`. Currently, on windows, which of these to use is based on whether or not dark mode for apps is enabled. However, the app dark mode setting is independent of the high contrast theme, which may be black or white. Now, check the high contrast theme to determine which `MaterialApp` theme to use when high contrast mode is enabled.

[flutter/flutter#75883](https://github.com/flutter/flutter/issues/75883)

*If you had to change anything in the [flutter/tests] repo, include a link to the migration guide as per the [breaking change policy].*

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt. See [testing the engine] for instructions on writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [ ] I signed the [CLA].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: shell/platform/windows/settings_plugin_unittests.cc [hunk 1]
```diff
@@ -70,5 +72,18 @@ TEST(SettingsPluginTest, StartWatchingStartsWatchingChanges) {
...
 }
 
+TEST(SettingsPluginTest, HighContrastModeHonored) {
+  TestBinaryMessenger messenger([](const std::string& channel,
+                                   const uint8_t* message, size_t message_size,
+                                   BinaryReply reply) {});
+  ::testing::NiceMock<MockSettingsPlugin> settings_plugin(&messenger, nullptr);
+
+  settings_plugin.UpdateHighContrastMode(true);
+  EXPECT_TRUE(settings_plugin.is_high_contrast());
```
Comment (loic-sharma) [inline shell/platform/windows/settings_plugin_unittests.cc:82]: Can we check this sends a message?

Comment (yaakovschectman) [reply shell/platform/windows/settings_plugin_unittests.cc:82]: I can make the unit test more robust. `SendSettingsSendsMessage` checks that the message is actually sent; I'll make sure `SendSettings` is called right.

File: shell/platform/windows/settings_plugin_unittests.cc [hunk 2]
```diff
@@ -70,5 +72,21 @@ TEST(SettingsPluginTest, StartWatchingStartsWatchingChanges) {
   settings_plugin.StartWatching();
 }
 
+TEST(SettingsPluginTest, HighContrastModeHonored) {
+  int times = 0;
+  TestBinaryMessenger messenger([&times](const std::string& channel,
+                                   const uint8_t* message, size_t message_size,
+                                   BinaryReply reply) { times++; });
```
Comment (loic-sharma) [inline shell/platform/windows/settings_plugin_unittests.cc:79]: Nit: consider asserting on the channel name since that's an easy spot check

**Review by loic-sharma — APPROVED** [review]
LGTM

File: shell/platform/windows/settings_plugin.cc [hunk 3]
```diff
@@ -103,19 +103,29 @@ float SettingsPlugin::GetTextScaleFactor() {
...
-  if (result == 0) {
-    return use_light_theme ? SettingsPlugin::PlatformBrightness::kLight
-                           : SettingsPlugin::PlatformBrightness::kDark;
+  if (is_high_contrast_) {
+    DWORD text_color = GetSysColor(COLOR_WINDOW);
+    int r = GetRValue(text_color);
+    int g = GetGValue(text_color);
+    int b = GetBValue(text_color);
+    int luminance = (r + r + r + b + (g << 2)) >> 3;
+    return luminance >= 0.5 ? SettingsPlugin::PlatformBrightness::kLight
```
Comment (cbracken) [inline shell/platform/windows/settings_plugin.cc:112]: Might be nice to extract this prologue to a `GetLuminance` static function like:
```cc
namespace {
int GetLuminance(DWORD color) {
  ...
}
}  // namespace
```

File: shell/platform/windows/settings_plugin.cc [hunk 4]
```diff
@@ -103,19 +103,29 @@ float SettingsPlugin::GetTextScaleFactor() {
...
+                            : SettingsPlugin::PlatformBrightness::kDark;
   } else {
-    // The current OS does not support dark mode. (Older Windows 10 or before
-    // Windows 10)
-    return SettingsPlugin::PlatformBrightness::kLight;
+    DWORD use_light_theme;
+    DWORD use_light_theme_size = sizeof(use_light_theme);
+    LONG result = RegGetValue(HKEY_CURRENT_USER, kGetPreferredBrightnessRegKey,
+                              kGetPreferredBrightnessRegValue, RRF_RT_REG_DWORD,
+                              nullptr, &use_light_theme, &use_light_theme_size);
```
Comment (cbracken) [inline shell/platform/windows/settings_plugin.cc:119]: Similarly here, it might be useful to extract this to a named function with a short doc comment.

**Review by cbracken — APPROVED** [review]
lgtm modulo minor nits.

LGTM stamp from a Japanese personal seal

**Pull request merged**